File: Java/src/com/daily/algothrim/sort/MergeSort.java
package com.daily.algothrim.sort;

/**
 * 归并排序
 * 稳定排序
 * 不是原地排序
 */
public class MergeSort {

    public static void main(String[] args) {
        int[] a = new int[]{10, 1, 4, 2, 5, 643, 654, 234, 2};
        new MergeSort().sort(a, 0, a.length - 1);
        for (int value : a) {
            System.out.println(value);
        }
    }

    /**
     * 时间复杂度: O(nlog n)
     * 空间复杂度: O(n)
     */
    private void sort(int[] a, int start, int end) {
        if (start < end) {
            int mid = (end + start) / 2;
            sort(a, start, mid);
            sort(a, mid + 1, end);
            merge(a, start, end);
        }
    }

    private void merge(int[] a, int start, int end) {
        int mid = (start + end) / 2;
        int i = start;
        int j = mid + 1;
        int k = 0;
        int[] temp = new int[end - start + 1];
        while (i <= mid && j <= end) {
            if (a[i] < a[j]) {
                temp[k++] = a[i++];
            } else {
                temp[k++] = a[j++];
            }
        }

        while (i <= mid) {
            temp[k++] = a[i++];
        }

        while (j <= end) {
            temp[k++] = a[j++];
        }

        if (temp.length >= 0) System.arraycopy(temp, 0, a, start, temp.length);
    }
}

File: kotlin/src/com/daily/algothrim/sort/MergeSort.kt
package com.daily.algothrim.sort


/**
 * 归并排序
 */
class MergeSort {
    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val a = intArrayOf(10, 1, 4, 2, 5, 643, 654, 234, 2)
            MergeSort().sort(a, 0, a.size - 1)
            a.forEach {
                println(it)
            }
        }
    }

    private fun sort(a: IntArray, start: Int, end: Int) {
        if (start < end) {
            val mid = (start + end) / 2
            // 分治递归
            sort(a, start, mid)
            sort(a, mid + 1, end)
            // 合并
            merge(a, start, end)
        }
    }

    private fun merge(a: IntArray, start: Int, end: Int) {
        val mid = (start + end) / 2
        var i = start
        var j = mid + 1
        var k = 0
        val temp = IntArray(end - start + 1)
        while (i <= mid && j <= end) {
            if (a[i] < a[j]) {
                temp[k++] = a[i++]
            } else {
                temp[k++] = a[j++]
            }
        }

        // i有剩余
        while (i <= mid) {
            temp[k++] = a[i++]
        }

        // j有剩余
        while (j <= end) {
            temp[k++] = a[j++]
        }

        // 将temp转移到a中
        temp.forEachIndexed { index, item ->
            a[start + index] = item
        }
    }
}